libs/SmartSync/src/com/salesforce/androidsdk/smartsync/util/SyncUpTarget.java

```diff
@@ -0,0 +1,200 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.smartsync.util;
+
+import android.text.TextUtils;
+import android.util.Log;
+
+import com.salesforce.androidsdk.rest.RestRequest;
+import com.salesforce.androidsdk.rest.RestResponse;
+import com.salesforce.androidsdk.smartstore.store.SmartStore;
+import com.salesforce.androidsdk.smartsync.manager.SyncManager;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.text.ParseException;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Target for sync up:
+ * - what records to upload to server
+ * - how to upload those records
+ */
+public class SyncUpTarget implements SyncTarget {
+
+    // Constants
+    public static final String ANDROID_IMPL = "androidImpl";
+
+    /**
+     * Build default SyncUpTarget
+     * @return
+     */
+    public static SyncUpTarget defaultSyncUpTarget() {
+        return new SyncUpTarget();
+    }
+
+    /**
+     * Build SyncUpTarget from json
+     *
+     * @param target as json
+     * @return
+     * @throws JSONException
+     */
+    @SuppressWarnings("unchecked")
+    public static SyncUpTarget fromJSON(JSONObject target) throws JSONException {
+        // Default sync up target
+        if (target == null || SyncUpTarget.class.getName().equals(target.getString(ANDROID_IMPL))) {
+            return SyncUpTarget.defaultSyncUpTarget();
+        }
+
+        // Non default sync up target
+        try {
+            Class<? extends SyncUpTarget> implClass = (Class<? extends SyncUpTarget>) Class.forName(target.getString(ANDROID_IMPL));
+            Method method = implClass.getMethod("fromJSON", JSONObject.class);
+            return (SyncUpTarget) method.invoke(null, target);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * @return json representation of target
+     * @throws JSONException
+     */
+    public JSONObject asJSON() throws JSONException {
+        JSONObject target = new JSONObject();
+        target.put(ANDROID_IMPL, getClass().getName());
+        return target;
+    }
+
+    /**
+     * Save locally created record back to server
+     * @param syncManager
+     * @param objectType
+     * @param fields
+     * @return server record id or null if creation failed
+     * @throws JSONException
+     * @throws IOException
+     */
+    public String createOnServer(SyncManager syncManager, String objectType, Map<String, Object> fields) throws JSONException, IOException {
```

Comment:
The logic to do a create/update/delete based on the dirty flag in the record remains in SyncManager. Every custom sync up target shouldn't have to implement it.